### PR TITLE
[IMP] mail: introduce message composition model

### PR DESCRIPTION
**Pull request by phenix-factory**
Task-2390637

### CODE REVIEW

File: addons/mail/static/src/models/composer_view.js [hunk 1]
```diff
@@ -673,7 +682,16 @@ registerModel({
                 this.updateMessage();
                 return;
             }
-            this.postMessage();
+            this.update({compositions: insert({
+                id: ev,
+                body: this._generateMessageBody()
+            })});
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:688]: style:
```js
this.update({
   compositions: insert({  })
})
```

File: addons/mail/static/src/models/composer_view.js [hunk 2]
```diff
@@ -673,7 +682,16 @@ registerModel({
                 this.updateMessage();
                 return;
             }
-            this.postMessage();
+            this.update({compositions: insert({
+                id: ev,
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:686]: better keep a counter locally and allocate a new one with `increment()`

File: addons/mail/static/src/models/composer_view.js [hunk 3]
```diff
@@ -1386,6 +1392,9 @@ registerModel({
             inverse: 'composerViewOwner',
             isCausal: true,
         }),
+        compositions: many('Composition', {
+            inverse: 'composerView',
+        }),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:1397]: `sort` by `id`, to avoid relying on arbitrary choice that no specified order here is FIFO due to implementation detail.

If this was LIFO instead, the behaviour would be bad for user, so makes sense to enforce order in modeling.

**Review by alexkuhn — COMMENTED** [review]
LGTM otherwise

File: addons/mail/static/src/models/composer_view.js [hunk 4]
```diff
@@ -633,11 +636,17 @@ registerModel({
                         threadView.update({ replyingToMessageView: clear() });
                     }
                 }
-                if (composer.exists()) {
-                    composer._reset();
+            }
+            catch {
+                err = true;
+            }
+            finally {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:643]: syntax:
```js
} catch (_err) {
} finally {
```

File: addons/mail/static/src/models/composer_view.js [hunk 5]
```diff
@@ -2,13 +2,21 @@
...
 import { escape, sprintf } from '@web/core/utils/strings';
 import { url } from '@web/core/utils/urls';
 
+const getComposerCompositionId = (function () {
+    let tmpId = 0;
+    return () => {
+        tmpId += 1;
+        return tmpId;
+    };
+})();
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:18]: This pattern sucks. Define a proper field and just `update({ increment() })` it

**Comment by alexkuhn** [comment]
also rename commit title so it's "composer composition" rather than "composition"

**Comment by seb-odoo** [comment]
Can we please name it `MessageComposition` instead? `ComposerComposition` means absolutely nothing to me.

File: addons/mail/static/src/models/composer_view.js [hunk 6]
```diff
@@ -676,31 +681,18 @@ registerModel({
          * Update a posted message when the message is ready.
          */
         async updateMessage() {
-            const composer = this.composer;
-            if (!composer.textInputContent) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:680]: looks unrelated diff?

File: addons/mail/static/src/models/composer_view.js [hunk 7]
```diff
@@ -676,31 +681,18 @@ registerModel({
...
-            body = parseAndTransform(body, addLink);
-            body = this._generateEmojisOnHtml(body);
             let data = {
-                body: body,
-                attachment_ids: composer.attachments.concat(this.messageViewInEditing.message.attachments).map(attachment => attachment.id),
+                body: this._generateMessageBody(),
+                attachment_ids: this.composer.attachments.concat(this.messageViewInEditing.message.attachments).map(attachment => attachment.id),
             };
-            try {
-                composer.update({ isPostingMessage: true });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:694]: looks unrelated diff?

File: addons/mail/static/src/models/composer_view.js [hunk 8]
```diff
@@ -676,31 +681,18 @@ registerModel({
...
-                const messageViewInEditing = this.messageViewInEditing;
-                await messageViewInEditing.message.updateContent(data);
-                if (messageViewInEditing.exists()) {
-                    messageViewInEditing.stopEditing();
-                }
-            } finally {
-                if (composer.exists()) {
-                    composer.update({ isPostingMessage: false });
-                }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:703]: looks unrelated diff?

File: addons/mail/static/src/models/composer_composition.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+/** @odoo-module **/
+
+import { registerModel } from '@mail/model/model_core';
+import { attr, one, many } from '@mail/model/model_field';
+
+registerModel({
+    name: 'ComposerComposition',
+    recordMethods: {
+        toParams() {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_composition.js:9]: looks like this should be a compute and not a method

File: addons/mail/static/src/models/composer_view.js [hunk 10]
```diff
@@ -535,38 +535,31 @@ registerModel({
...
-                'thread_model': composer.thread.model,
-            };
+            const composition = this.composition;
             try {
                 composer.update({ isPostingMessage: true });
                 if (composer.thread.model === 'mail.channel') {
-                    Object.assign(postData, {
-                        subtype_xmlid: 'mail.mt_comment',
-                    });
+                    composition.update({ subtype_xmlid: 'mail.mt_comment' });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:542]: too much diff, why not just have `const params = composition.params` above

File: addons/mail/static/src/models/composer_composition.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    },
+    fields: {
+        attachments: many('Attachment'),
+        id: attr({
+            identifying: true,
+        }),
+        chatter: one('Chatter'),
+        composerView: one('ComposerView', {
+            inverse: 'composition',
+        }),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_composition.js:32]: remove explicit inverse if not used

File: addons/mail/static/src/models/composer_view.js [hunk 12]
```diff
@@ -1159,7 +1151,7 @@ registerModel({
          * @private
          * @returns {Object}
          */
-        _getMessageData() {
+        _generateMessageBody() {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:1154]: split to another PR

File: addons/mail/static/src/models/composer_view.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/composer_view.js:680]: split to another PR

File: addons/mail/static/src/models/composer_view.js (hunk 10 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/composer_view.js:542]: -> compute

File: addons/mail/static/src/models/composer_view.js [hunk 13]
```diff
@@ -663,6 +644,18 @@ registerModel({
                 }
                 this.composer.thread.unregisterCurrentPartnerIsTyping({ immediateNotify: true });
             }
+            this.update({
+                composition: {
+                    attachments: this.composer.attachments,
+                    id: increment(),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:650]: This is not how `increment` works, it's meant to update the existing value of a field.

Comment (seb-odoo) [reply addons/mail/static/src/models/composer_view.js:650]: The counter needs to be stored separately, it can't be the id itself.

**Comment by phenix-factory** [comment]
@seb-odoo @alexkuhn This is rebased on the splited PR and I have build compute.
I also use `composer.activeThread` in the composition instead of `composer.thread`, it's needed to compute the `context` field.

File: addons/mail/static/src/models/composer_view.js [hunk 14]
```diff
@@ -1367,6 +1346,10 @@ registerModel({
             inverse: 'composerViewOwner',
             isCausal: true,
         }),
+        compositions: many('MessageComposition'),
+        compositionsIndex: attr({
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer_view.js:1350]: this value should be global, not limited to the current composer view

Comment (phenix-factory) [reply addons/mail/static/src/models/composer_view.js]: It's moved to messaging now
